src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,210 @@
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunkId, chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSStreamReference jsStreamReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            // Enforce minimum 1 kb SignalR message size as we budget 512 bytes
+            // overhead for the transfer, thus leaving at least 512 bytes for data
+            // transfer per chunk.
+            var chunkSize = maximumIncomingBytes > 1024 ?
+                maximumIncomingBytes - 512 :
+                throw new ArgumentException($"SignalR MaximumIncomingBytes must be at least 1 kb.");
+
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsStreamReference, streamId, chunkSize);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            long streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _streamCancellationToken = cancellationToken;
+
+            // Dispose of the stream if a chunk isn't received within 1 minute.
+            _receiveDataTimer = new Timer(new TimerCallback(async (state) => {
+                var stream = state as RemoteJSDataStream;
+                var timeoutException = new TimeoutException("Did not receive any data in the alloted time.");
+                await stream.CompletePipeAndDisposeStream(timeoutException);
+            }), this, dueTime: ReceiveDataTimeout, period: Timeout.InfiniteTimeSpan);
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize / 2));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // internal for testing
+        internal TimeSpan ReceiveDataTimeout { private get; set; } = TimeSpan.FromMinutes(1);
+
+        private async Task<bool> ReceiveData(long chunkId, byte[] chunk, string error)
+        {
+            try
+            {
+                // Reset the timeout as a chunk has been received
+                _receiveDataTimer.Change(dueTime: ReceiveDataTimeout, period: Timeout.InfiniteTimeSpan);
```

Comment:
```
var lastReceiveTime = DateTime.UtcNow;
_ = ThrowOnTimeout(lastReceiveTime);

async Task ThrowOnTimeout(DateTimeOffset lastReceivedTime)
{
    await Task.Delay(circuitOptions.JSInteropTimeout);
    var lastUpdate = _lastReceivedTime;
    if(lastUpdate >= lastReceivedTime.Add(circuitOptions.JSInteropTimeout))
    {
       // Dispatch error
    }
}
```

Reply:
Took a slightly different approach (https://github.com/dotnet/aspnetcore/pull/33491/commits/b5615c70578000c38c31e53bb1fda67ce7b70df3) based on cancellation tokens. This was to avoid having a few thousand pending tasks (possible for a large transfer over a minute) being awaited.